ros-safety/software_watchdogs
# Example launch files for dockerized watchdog & processes

This introduces the _"dockerized"_ watchdog and monitored process example introduced in the Autoware workshop at IV 2020. 

![dockerized](https://user-images.githubusercontent.com/478855/96934900-0adcfb80-1491-11eb-9c55-8aa93767bacf.jpg)

The choice of Docker for this example is because it facilitates use of cgroups (kernel control groups for resource allocation and isolation), and kernel namespaces for process isolation. The monitored entity can crash or be otherwise stopped/stalled and the Watchdog will respawn the entity.

To test the functionality in this PR:
- [Install docker](https://docs.docker.com/get-docker/) for your Linux distribution
- Check out this branch and edit the included Dockerfile as follows:
```
diff --git a/Dockerfile b/Dockerfile
index d6fa9fd..c3b79af 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -6,7 +6,7 @@ LABEL stage=intermediate
 
 # Clone and build repository
 WORKDIR /root
-RUN git clone https://github.com/ros-safety/software_watchdogs.git
+RUN git clone -b experimental-relaunch https://github.com/ros-safety/software_watchdogs.git
 RUN . /opt/ros/foxy/setup.sh && \
     colcon build
```
This will make sure that the correct branch is used in the next step (as long as this PR is not merged to master yet).
- Build the Docker image:
```
docker build --no-cache . -t sw_watchdogs:latest
```
- Run the demo (`watchdog_in_docker.launch.py`) via:
```
docker run -it --rm --name testing -v /var/run/docker.sock:/var/run/docker.sock \
-v /usr/bin/docker:/usr/bin/docker sw_watchdogs:latest ros2 launch sw_watchdog watchdog_in_docker.launch.py
```
This command brings up both the watchdog and the monitored application in two separate containers. See `watchdog_in_docker.launch.py` for details. Expect terminal output as follows after startup:

<img src="https://user-images.githubusercontent.com/478855/96937031-424da700-1495-11eb-967e-78862be881a7.jpg" alt="terminal" width="75%" height="75%">

- Manually interfere with the monitored application, for example via: 
```
docker kill --signal=SIGINT talker
```
This will kill the monitored entity (called `talker`) and force the watchdog to restart it shortly after. In the original terminal, this restart activity will be reflected as follows:

<img src="https://user-images.githubusercontent.com/478855/96937283-c869ed80-1495-11eb-952a-97b0491d2f9a.jpg" alt="terminal-2" width="75%" height="75%">
